service/src/test/java/bio/terra/workspace/service/resource/controlled/flight/CreateAzureIpStepTest.java

```diff
@@ -0,0 +1,96 @@
+package bio.terra.workspace.service.resource.controlled.flight;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.when;
+
+import bio.terra.cloudres.azure.resourcemanager.compute.ComputeManagerCow;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.StepResult;
+import bio.terra.workspace.app.configuration.external.AzureConfiguration;
+import bio.terra.workspace.app.configuration.external.AzureState;
+import bio.terra.workspace.common.BaseUnitTest;
+import bio.terra.workspace.common.fixtures.ControlledResourceFixtures;
+import bio.terra.workspace.generated.model.ApiAzureIpCreationParameters;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.flight.create.CreateAzureIpStep;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.core.management.Region;
+import com.azure.core.util.Context;
+import com.azure.resourcemanager.compute.ComputeManager;
+import com.azure.resourcemanager.network.NetworkManager;
+import com.azure.resourcemanager.network.models.PublicIpAddress;
+import com.azure.resourcemanager.network.models.PublicIpAddresses;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ActiveProfiles;
+
+@ActiveProfiles("azure")
+public class CreateAzureIpStepTest extends BaseUnitTest {
+
+  @Mock private FlightContext mockFlightContext;
+  @Mock private CrlService mockCrlService;
+  @Mock private ComputeManagerCow mockComputeCow;
+  @Mock private WorkspaceService mockWorkspaceService;
+  @Mock private AzureConfiguration mockAzureConfig;
+  @Mock private AzureCloudContext mockAzureCloudContext;
+  @Mock private PublicIpAddress mockPublicIpAddress;
+  @Mock private ComputeManager mockComputeManager;
+  @Mock private NetworkManager mockNetworkManager;
+  @Mock private PublicIpAddresses mockPublicIpAddresses;
+  @Mock private PublicIpAddress.DefinitionStages.Blank mockIpStage1;
+  @Mock private PublicIpAddress.DefinitionStages.WithGroup mockIpStage2;
+  @Mock private PublicIpAddress.DefinitionStages.WithCreate mockIpStage3;
+  private static final String STUB_STRING_RETURN = "stubbed-return";
+  @Mock private Region mockRegion;
+
+  @Autowired private AzureState azureState;
+
+  @BeforeEach
+  public void setup() {
+    when(mockAzureCloudContext.getAzureResourceGroupId()).thenReturn(STUB_STRING_RETURN);
```

Comment:
this was fun.... mockito theoretically supports chain-stubbing, but it didn't seem like it cleaned up the syntax too much on cursory investigation.